repo_policy_compliance/blueprint.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provides API blueprint for flask to run the policy checks."""
+
+import logging
+import os
+import secrets
+from typing import cast
+
+from flask import Blueprint, Response, request
+from flask_httpauth import HTTPTokenAuth
+
+from . import Result, all_
+
+repo_policy_compliance = Blueprint("repo_policy_compliance", __name__)
+auth = HTTPTokenAuth(scheme="Bearer")
+runner_tokens: set[str] = set()
+
+# Bandit thinks this is the token value when it is the name of the environment variable with the
+# token value
+CHARM_TOKEN_ENV_NAME = "CHARM_TOKEN"  # nosec
+# Bandit thinks this is the token value when it is the name of the endpoint to get a one time token
+ONE_TIME_TOKEN_ENDPOINT = "/one-time-token"  # nosec
+CHECK_RUN_ENDPOINT = "/check-run"
+CHARM_USER = "charm"
+CHARM_ROLE = CHARM_USER
+RUNNER_USER = "runner"
+RUNNER_ROLE = RUNNER_USER
+
+EXPECTED_KEYS = {
+    "repository_name",
+    "source_repository_name",
+    "target_branch_name",
+    "source_branch_name",
+    "commit_sha",
+}
+
+
+@auth.verify_token
+def verify_token(token: str) -> str | None:
+    """Verify the authentication token.
+
+    Args:
+        token: The token to check.
+
+    Returns:
+        The identity associated with the token or None if no token matches.
+    """
+    charm_token = os.getenv(CHARM_TOKEN_ENV_NAME)
+
+    if not charm_token:
+        logging.error(
+            (
+                "%s environment variable is required for generating one time tokens, it is not "
+                "defined or empty"
+            ),
+            CHARM_TOKEN_ENV_NAME,
+        )
+        return None
+
+    if token == charm_token:
+        return CHARM_USER
+
+    if token in runner_tokens:
+        runner_tokens.remove(token)
```

Comment:
Flask runs in a multi-thread environment, I think the step here might not be thread safe.

Reply:
Yes this is not threat safe. For simplicity, the application will be run with a single worker on each unit which should not overload the single worker. The alternative is to use a fully fledged database which would increase the complexity of the solution by quite a lot, unless you can think of any other ways to make this thread safe without using a database?

Reply:
There's no way to share data across processes without the configuration on the wsgi server. But there's no control in Flask to prevent a multi-thread multi-process deployment. Maybe we should state that in the docstring at lease.

Reply:
Good point, added to the module docstring